src/main/java/com/google/api/codegen/gapic/GapicGeneratorFactory.java

```diff
@@ -74,97 +78,93 @@
 import java.util.Arrays;
 import java.util.List;
+import java.util.function.Function;
 
 /** GapicGeneratorFactory creates CodeGenerator instances based on an id. */
 public class GapicGeneratorFactory {
-  public static final String ARTIFACT_SURFACE = "surface";
-  public static final String ARTIFACT_TEST = "test";
-  public static final String ARTIFACT_SAMPLES = "samples";
-
-  public static boolean enableSurfaceGenerator(List<String> enabledArtifacts) {
-    return enabledArtifacts.isEmpty() || enabledArtifacts.contains(ARTIFACT_SURFACE);
-  }
-
-  public static boolean enableTestGenerator(List<String> enabledArtifacts) {
-    return enabledArtifacts.isEmpty() || enabledArtifacts.contains(ARTIFACT_TEST);
-  }
-
-  public static boolean enableSampleGenerator(List<String> enabledArtifacts) {
-    return enabledArtifacts.contains(ARTIFACT_SAMPLES);
-  }
 
   /** Create the GapicGenerators based on the given id */
   public static List<CodeGenerator<?>> create(
       TargetLanguage language,
       Model model,
       GapicProductConfig productConfig,
       PackageMetadataConfig packageConfig,
-      List<String> enabledArtifacts) {
+      ArtifactFlags artifactFlags) {
 
     ArrayList<CodeGenerator<?>> generators = new ArrayList<>();
     // Please keep the following IDs in alphabetical order
     if (language.equals(CSHARP)) {
       String packageName = productConfig.getPackageName();
-      if (enableSurfaceGenerator(enabledArtifacts)) {
-        GapicCodePathMapper pathMapper =
-            CommonGapicCodePathMapper.newBuilder()
-                .setPrefix(packageName + File.separator + packageName)
-                .setPackageFilePathNameFormatter(new CSharpNameFormatter())
-                .build();
-        CodeGenerator mainGenerator =
-            GapicGenerator.newBuilder()
-                .setModel(model)
-                .setProductConfig(productConfig)
-                .setSnippetSetRunner(new CommonSnippetSetRunner(new CSharpRenderingUtil()))
-                .setModelToViewTransformer(
-                    new CSharpGapicClientTransformer(pathMapper, packageConfig))
-                .build();
-        generators.add(mainGenerator);
-        GapicCodePathMapper snippetPathMapper =
-            CommonGapicCodePathMapper.newBuilder()
-                .setPrefix(packageName + File.separator + packageName + ".Snippets")
-                .setPackageFilePathNameFormatter(new CSharpNameFormatter())
-                .build();
-        CodeGenerator snippetGenerator =
-            GapicGenerator.newBuilder()
-                .setModel(model)
-                .setProductConfig(productConfig)
-                .setSnippetSetRunner(new CommonSnippetSetRunner(new CSharpRenderingUtil()))
-                .setModelToViewTransformer(new CSharpGapicSnippetsTransformer(snippetPathMapper))
-                .build();
-        generators.add(snippetGenerator);
+
+      Function<String, GapicCodePathMapper> newCodePathMapper =
```

Comment:
They are lambda variables because that creates the most concise, non-repetitive code. If I created private methods for each of these, they would have about as many parameters as they had lines of code, and the resulting code would actually be the same size as the original, which would defeat the purpose. I feel that lambdas are highly appropriate here because they can reference the local scope, and be scoped to that local scope at the same time.

Reply:
Good point. I didn't notice the local references in the lambdas. `newCodePathMapper` appears to only have 2 local references though. Up to you if it's worth moving to a private method.

Reply:
newCodePathMapper has four local references; I'd prefer to keep it a lambda because the logic can be kept closer to where it is used (`create()` is a mega-function).